### PR TITLE
Remove auth data from log text string

### PR DESCRIPTION
**Pull request by JoseSanchez7**
**Description of the problem:** 
text_line gets sent to the wire with sensitive information and is therefore logging info such as SAS tokens and proxy authentication info.

**Description of the the solution:**
Pass the text_line string with the following headers removed:

Authorization
Proxy-Authorization

### CODE REVIEW

File: src/uhttp.c [hunk 1]
```diff
@@ -310,19 +310,59 @@ static int process_header_line(const unsigned char* buffer, size_t len, size_t*
...
+
+            else if (proxyAuthStart != NULL && authStart != NULL)
+            {
+                char* proxyAuthEol = strstr(proxyAuthStart, "\x0D\x0A");
+
+                if (proxyAuthEol != NULL)
+                {
+                    LOG(AZ_LOG_TRACE, LOG_LINE, "%.*s **************\n************** %.*s", (int)(authStart  - text_line), text_line, (int)(strlen(text_line) - (proxyAuthEol - text_line)), proxyAuthEol);
+                }
+            }
```
[the remark this answers is not in the document]
Comment (JoseSanchez7) [reply src/uhttp.c]: How would they get shuffled around? It's being passed as one string.

File: src/uhttp.c [hunk 2]
```diff
@@ -310,19 +310,59 @@ static int process_header_line(const unsigned char* buffer, size_t len, size_t*
...
+
+                if (proxyAuthEol != NULL)
+                {
+                    LOG(AZ_LOG_TRACE, LOG_LINE, "%.*s ************** %.*s", (int)(proxyAuthStart  - text_line), text_line, (int)(strlen(text_line) - (proxyAuthEol - text_line)), proxyAuthEol);
+                }
+            }
+
+            else if (proxyAuthStart != NULL && authStart != NULL)
+            {
+                char* proxyAuthEol = strstr(proxyAuthStart, "\x0D\x0A");
```
[the remark this answers is not in the document]
Comment (JoseSanchez7) [reply src/uhttp.c:348]: It seems it will as that is how the json is formatted.

File: src/uhttp.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (danewalton) [reply src/uhttp.c]: The header is constructed somewhere, and if the ordering of that construction gets changed, then auth start might come before proxy auth start, and you'll miss one of the auth's and then redact something that isn't related to auth.

Essentially, you can't assume here the ordering of proxy auth and auth. You have to redact them separately.

File: src/uhttp.c [hunk 3]
```diff
@@ -310,19 +310,59 @@ static int process_header_line(const unsigned char* buffer, size_t len, size_t*
...
     }
+
     else
     {
         result = 0;
+
         if (http_data->trace_on)
         {
-            LOG(AZ_LOG_TRACE, LOG_LINE, "%s", text_line);
+            char* authStart = strstr(text_line, "Authorization:");
```
[the remark this answers is not in the document]
Comment (danewalton) [reply src/uhttp.c:324]: sorry I should have clarified. Any string literal here, like `"Authorization:"` or `"\x0D\x0A"` should be declared at the top of the file as a `static const char*` and then that variable should be used in place of the values here and elsewhere in your changes. Take a look at the top of this file for other values that are declared and used that way.

File: src/uhttp.c [hunk 4]
```diff
@@ -320,7 +321,36 @@ static int write_text_line(HTTP_CLIENT_HANDLE_DATA* http_data, const char* text_
...
+            char* proxyAuthStart = strstr(text_line, "Proxy-Authorization");
+            if (authStart == NULL && proxyAuthStart == NULL)
+            {
+                LOG(AZ_LOG_TRACE, LOG_LINE, "%s", text_line);
+            }
+            else if (authStart != NULL && proxyAuthStart == NULL)
+            {
+                char* authEol = strstr(authStart, "\x0D\x0A");
+                if (authEol != NULL)
```
Comment (vaavva) [inline src/uhttp.c:333]: If this is null, should we still print up to the authStart or something like that, or does this mean that the `text_line`  has something wrong with it so it's not worth printing?

Comment (vaavva) [reply src/uhttp.c:333]: (same question for line 341)

File: src/uhttp.c [hunk 5]
```diff
@@ -320,7 +321,36 @@ static int write_text_line(HTTP_CLIENT_HANDLE_DATA* http_data, const char* text_
...
         if (http_data->trace_on)
         {
-            LOG(AZ_LOG_TRACE, LOG_LINE, "%s", text_line);
+            char* authStart = strstr(text_line, "Authorization:");
+            char* proxyAuthStart = strstr(text_line, "Proxy-Authorization");
+            if (authStart == NULL && proxyAuthStart == NULL)
+            {
+                LOG(AZ_LOG_TRACE, LOG_LINE, "%s", text_line);
+            }
+            else if (authStart != NULL && proxyAuthStart == NULL)
```
[the remark this answers is not in the document]
Comment (ericwol-msft) [reply src/uhttp.c:330]: No, there is no requirement that they use proxy that require authentication. Same is true for TLS cert auth, so "Authorization:" is also optional in the HTTP header.

File: src/uhttp.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (ericwol-msft) [reply src/uhttp.c:348]: JSON?? This is the HTTP header.

File: src/uhttp.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (vaavva) [reply src/uhttp.c:330]: My comment was geared towards this line being an else if after this if statement `(authStart == NULL && proxyAuthStart == NULL)` that if both aren't null, then in any case where one is null, the other must not be null as well, so this check could be simplified to  else if proxyAuthStart == NULL

File: src/uhttp.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (ericwol-msft) [reply src/uhttp.c:348]: Header fields are colon-separated key-value pairs in clear-text [string](https://en.wikipedia.org/wiki/String_(computer_science)) format, terminated by a [carriage return](https://en.wikipedia.org/wiki/Carriage_return) (CR) and [line feed](https://en.wikipedia.org/wiki/Line_feed) (LF) character sequence. The end of the header section is indicated by an empty field line, resulting in the transmission of two consecutive CR-LF pairs.

https://en.wikipedia.org/wiki/List_of_HTTP_header_fields

[the remark this answers is not in the document]
Comment (ericwol-msft) [reply src/uhttp.c:348]: Could be safe and just look for \x0A

File: src/uhttp.c [hunk 6]
```diff
@@ -320,7 +321,36 @@ static int write_text_line(HTTP_CLIENT_HANDLE_DATA* http_data, const char* text_
         result = 0;
         if (http_data->trace_on)
         {
-            LOG(AZ_LOG_TRACE, LOG_LINE, "%s", text_line);
+            char* authStart = strstr(text_line, "Authorization:");
+            char* proxyAuthStart = strstr(text_line, "Proxy-Authorization");
```
Comment (ericwol-msft) [inline src/uhttp.c:325]:
```suggestion
            char* proxyAuthStart = strstr(text_line, "Proxy-Authorization:");
```

File: src/uhttp.c (hunk 4 above)
Comment (ericwol-msft) [reply src/uhttp.c:333]: I'm fine not printing it. mal-formed HTML, should not happen, this is only trace code

File: src/uhttp.c [hunk 7]
```diff
@@ -320,7 +321,36 @@ static int write_text_line(HTTP_CLIENT_HANDLE_DATA* http_data, const char* text_
...
+                {
+                    LOG(AZ_LOG_TRACE, LOG_LINE, "%.*s ************** %.*s", (int)(proxyAuthStart  - text_line), text_line, (int)(strlen(text_line) - (proxyAuthEol - text_line)), proxyAuthEol);
+                }
+            }
+            else if (proxyAuthStart != NULL && authStart != NULL)
+            {
+                char* proxyAuthEol = strstr(proxyAuthStart, "\x0D\x0A");
+                if (proxyAuthEol != NULL)
+                {
+                    LOG(AZ_LOG_TRACE, LOG_LINE, "%.*s **************\n************** %.*s", (int)(authStart  - text_line), text_line, (int)(strlen(text_line) - (proxyAuthEol - text_line)), proxyAuthEol);
```
Comment (ericwol-msft) [inline src/uhttp.c:351]: ???
There is no position order in the header of the fields, so authStart  > proxyAuthStart

Comment (ericwol-msft) [reply src/uhttp.c:351]: need to handle each span separately and figure out which one comes first.

File: src/uhttp.c [hunk 8]
```diff
@@ -320,7 +325,48 @@ static int write_text_line(HTTP_CLIENT_HANDLE_DATA* http_data, const char* text_
...
+                if (authEol != NULL)
+                {                                                                 
+                    LOG(AZ_LOG_TRACE, LOG_LINE, "%.*s ************** %.*s", (int)(authStart  - text_line), text_line, (int)(strlen(text_line) - (authEol - text_line)), authEol);
+                }
+            }
+            else if (proxyAuthStart != NULL && authStart == NULL)
+            {
+                if (proxyAuthEol != NULL)
+                {
+                    LOG(AZ_LOG_TRACE, LOG_LINE, "%.*s ************** %.*s", (int)(proxyAuthStart  - text_line), text_line, (int)(strlen(text_line) - (proxyAuthEol - text_line)), proxyAuthEol);
```
Comment (ericwol-msft) [inline src/uhttp.c:347]: what are the ************** different length in each log log line?

File: src/uhttp.c [hunk 9]
```diff
@@ -320,7 +325,48 @@ static int write_text_line(HTTP_CLIENT_HANDLE_DATA* http_data, const char* text_
         result = 0;
         if (http_data->trace_on)
         {
-            LOG(AZ_LOG_TRACE, LOG_LINE, "%s", text_line);
+            char* authStart = strstr(text_line, AUTH_HEADER);
+            char* proxyAuthStart = strstr(text_line, PROXY_AUTH_HEADER);
+            char* authEol = strstr(authStart, HEADER_ENDING);
```
Comment (ericwol-msft) [inline src/uhttp.c:330]: What happen if AUTH_HEADER or PROXY_AUTH_HEADER is not in the header? you are passing null to strstr()

The ISO C standard says that the behavior is undefined.

File: src/uhttp.c [hunk 10]
```diff
@@ -320,7 +325,48 @@ static int write_text_line(HTTP_CLIENT_HANDLE_DATA* http_data, const char* text_
...
+                int proxyAuthPos = proxyAuthStart - text_line;
+                if(authPos < proxyAuthPos)
+                {
+                    LOG(AZ_LOG_TRACE, LOG_LINE, "%.*s ***** %.*s ***** %.*s", (int)(authStart - text_line), text_line, (int)(proxyAuthStart - authEol), authEol,
+(int)(proxyAuthEol - text_line),
+proxyAuthEol);
+                }
+                else
+                {
+                    char* secondAuthStart = strstr(text_line, AUTH_HEADER_TWO);
```
Comment (ericwol-msft) [inline src/uhttp.c:362]: why this search again? you already did that at the start. just swap how the output is formatted

File: src/uhttp.c [hunk 11]
```diff
@@ -32,6 +32,10 @@ static const char* HTTP_HOST = "Host";
 static const char* HTTP_CONTENT_LEN = "content-length";
 static const char* HTTP_TRANSFER_ENCODING = "transfer-encoding";
 static const char* HTTP_CRLF_VALUE = "\r\n";
+static const char* HEADER_ENDING = "\x0A";
+static const char* AUTH_HEADER = "Authorization:";
+static const char* AUTH_HEADER_TWO = "\nAuthorization";
```
Comment (vaavva) [inline src/uhttp.c:37]:
```suggestion
static const char* AUTH_HEADER_TWO = "\nAuthorization:";
```
should this have the colon as well?

File: src/uhttp.c [hunk 12]
```diff
@@ -320,7 +325,48 @@ static int write_text_line(HTTP_CLIENT_HANDLE_DATA* http_data, const char* text_
         result = 0;
         if (http_data->trace_on)
         {
-            LOG(AZ_LOG_TRACE, LOG_LINE, "%s", text_line);
+            char* authStart = strstr(text_line, AUTH_HEADER);
+            char* proxyAuthStart = strstr(text_line, PROXY_AUTH_HEADER);
+            char* authEol = strstr(authStart, HEADER_ENDING);
+            char* proxyAuthEol = strstr(proxyAuthStart, HEADER_ENDING);
+            if (authStart == NULL && proxyAuthStart == NULL)
```
Comment (vaavva) [inline src/uhttp.c:332]: Can you add unit tests for these 5 scenarios since there's a lot of string manipulation going on? (both null, only proxy null, only auth null, neither null with proxy first, and neither null with proxy second)

File: src/uhttp.c [hunk 13]
```diff
@@ -320,7 +325,48 @@ static int write_text_line(HTTP_CLIENT_HANDLE_DATA* http_data, const char* text_
...
+                LOG(AZ_LOG_TRACE, LOG_LINE, "%s", text_line);
+            }
+            else if (authStart != NULL && proxyAuthStart == NULL)
+            {
+                if (authEol != NULL)
+                {                                                                 
+                    LOG(AZ_LOG_TRACE, LOG_LINE, "%.*s ************** %.*s", (int)(authStart  - text_line), text_line, (int)(strlen(text_line) - (authEol - text_line)), authEol);
+                }
+            }
+            else if (proxyAuthStart != NULL && authStart == NULL)
```
Comment (vaavva) [inline src/uhttp.c:343]: Since `Authorization:` is in `Proxy-Authorization:`, if proxyAuthStart != NULL, won't authStart always != NULL?

File: src/uhttp.c (hunk 10 above)
Comment (JoseSanchez7) [reply src/uhttp.c:362]: Because the first "Authorization" is picked up in "Proxy-Authorization:", and it messes things up. That's why I search for it with AUTH_HEADER_TWO, which includes a '\n' at the beginning.

File: src/uhttp.c (hunk 13 above)
Comment (JoseSanchez7) [reply src/uhttp.c:343]: Great catch, I will fix that.

File: src/uhttp.c [hunk 14]
```diff
@@ -307,9 +310,63 @@ static int process_header_line(const unsigned char* buffer, size_t len, size_t*
     return result;
 }
 
+static void log_data_line(HTTP_CLIENT_HANDLE_DATA* http_data, const char* text_line)
+{
+    char* authStart = strstr(text_line, AUTH_HEADER);
```
Comment (ewertons) [inline src/uhttp.c:315]: So here, if Proxy-Authorization is before the Authorization header, authStart could be set to the middle of Proxy-Authorization. That is not the correct logic.

File: src/uhttp.c [hunk 15]
```diff
@@ -307,9 +310,63 @@ static int process_header_line(const unsigned char* buffer, size_t len, size_t*
...
+    char* proxyAuthEol = NULL;
+
+    if (authStart != NULL)
+    {
+        authEol = strstr(authStart, HEADER_ENDING);
+    }
+    if (proxyAuthStart != NULL)
+    {
+        proxyAuthEol = strstr(proxyAuthStart, HEADER_ENDING);
+        authStart = strstr(proxyAuthEol, AUTH_HEADER);
```
Comment (ewertons) [inline src/uhttp.c:327]: So here if proxyAuthStart is not NULL, you are overriding authStart?
In that case the previous code that sets or checks authStart is moot.

**Review by ewertons — CHANGES_REQUESTED** [review]

**Comment by ewertons** [comment]
Here I think the logic is not well defined to fix this issue.

This goal:
log_data_line() receives `text_line`, which contains the complete HTTP request.
It must then log that HTTP request, obfuscating the values of headers Authorization and Proxy-Authorization, if they are defined.

Here is a suggestion of a solution:
Search for \r\nAuthorization\r\n, store it in `authStart`.
Search for \r\nProxy-Authorization\r\n, store it in `proxyAuthStart`.
If both headers are not present, print `text_line` as is and return from the function.
Otherwise, 
Set both `authEnd` and `proxyAuthEnd` (to the beginning or the following CRLF) using the respective `authStart` (+1) and `proxyAuthStart` (+1), if they are present.
Log `text_line` from position 0 to the lesser of `authStart` and `proxyAuthStart`.
Log "\r\nAuthorization: ***" and "\r\nProxy-Authorization: ***".
Log `text_line` from `authEnd` or `proxyAuthEnd` (whichever is lower) up to `authStart` or `proxyAuthStart` (whichever is higher) (be careful here to also check if Authorization and Proxy-Authorization are next to each other).
Log `text_line` from `authEnd` or `proxyAuthEnd` (whichever is higher) to the end of `text_line`.

File: src/uhttp.c [hunk 16]
```diff
@@ -307,9 +309,58 @@ static int process_header_line(const unsigned char* buffer, size_t len, size_t*
...
+    else if (proxyAuthStart != NULL && authStart == NULL)
+    {
+        if (proxyAuthEol != NULL)
+        {
+            LOG(AZ_LOG_TRACE, LOG_LINE, "%.*s \r\nProxy-Authorization: *** %.*s", (int)(proxyAuthStart  - text_line), text_line, (int)(strlen(text_line) - (proxyAuthEol - text_line)), proxyAuthEol);
+        }
+    }
+    else
+    {
+        long int authPos = authStart - text_line;
```
Comment (ewertons) [inline src/uhttp.c:347]: Why do you need authPos and proxyAuthPos? Can't you just use authStart and proxyAuthStart in the comparison on line 349?

**Review by ewertons — CHANGES_REQUESTED** [review]

**Review by ewertons — APPROVED** [review]

File: src/uhttp.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (JoseSanchez7) [reply src/uhttp.c]: Thanks @danewalton ! I went ahead and fixed that.

File: src/uhttp.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (JoseSanchez7) [reply src/uhttp.c]: Advice taken. Thanks!

File: src/uhttp.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (JoseSanchez7) [reply src/uhttp.c]: I went ahead and stuck with "\x0D\x0A" as the endings, as well as the preceding characters for each header, since that is what is described here: https://www.rfc-editor.org/rfc/rfc2616#section-5.

File: src/uhttp.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (JoseSanchez7) [reply src/uhttp.c]: Thanks for pointing that out Val! I went ahead and fixed it.

File: src/uhttp.c [hunk 17]
```diff
@@ -320,7 +321,36 @@ static int write_text_line(HTTP_CLIENT_HANDLE_DATA* http_data, const char* text_
...
+            }
+            else if (proxyAuthStart != NULL && authStart == NULL)
+            {
+                char* proxyAuthEol = strstr(proxyAuthStart, "\x0D\x0A");
+                if (proxyAuthEol != NULL)
+                {
+                    LOG(AZ_LOG_TRACE, LOG_LINE, "%.*s ************** %.*s", (int)(proxyAuthStart  - text_line), text_line, (int)(strlen(text_line) - (proxyAuthEol - text_line)), proxyAuthEol);
+                }
+            }
+            else if (proxyAuthStart != NULL && authStart != NULL)
```
[the remark this answers is not in the document]
Comment (JoseSanchez7) [reply src/uhttp.c]: Fixed.

File: src/uhttp.c (hunk 4 above)
Comment (JoseSanchez7) [reply src/uhttp.c]: I'm only printing for well formed HTML.

File: src/uhttp.c (hunk 6 above)
Comment (JoseSanchez7) [reply src/uhttp.c]: Suggestion committed.

File: src/uhttp.c (hunk 7 above)
Comment (JoseSanchez7) [reply src/uhttp.c]: Fixed this. Thanks!

File: src/uhttp.c (hunk 8 above)
Comment (JoseSanchez7) [reply src/uhttp.c]: Good point. Fixed.

File: src/uhttp.c (hunk 9 above)
Comment (JoseSanchez7) [reply src/uhttp.c]: Fixed.

File: src/uhttp.c (hunk 11 above)
Comment (JoseSanchez7) [reply src/uhttp.c]: Good point! Fixed.

File: src/uhttp.c (hunk 12 above)
Comment (JoseSanchez7) [reply src/uhttp.c]: I won't be adding unit test since this is related to logging (per Ewerton's advice). But I am doing individual test for each scenario with a sample.

File: src/uhttp.c [hunk 18]
```diff
@@ -32,6 +32,9 @@ static const char* HTTP_HOST = "Host";
 static const char* HTTP_CONTENT_LEN = "content-length";
 static const char* HTTP_TRANSFER_ENCODING = "transfer-encoding";
 static const char* HTTP_CRLF_VALUE = "\r\n";
+static const char* HEADER_ENDING = "\x0A";
```
[the remark this answers is not in the document]
Comment (JoseSanchez7) [reply src/uhttp.c]: Good point, thanks!

File: src/uhttp.c [hunk 19]
```diff
@@ -320,7 +377,7 @@ static int write_text_line(HTTP_CLIENT_HANDLE_DATA* http_data, const char* text_
         result = 0;
         if (http_data->trace_on)
         {
-            LOG(AZ_LOG_TRACE, LOG_LINE, "%s", text_line);
+            log_data_line;
```
[the remark this answers is not in the document]
Comment (JoseSanchez7) [reply src/uhttp.c]: Tested locally.

File: src/uhttp.c [hunk 20]
```diff
@@ -307,9 +310,63 @@ static int process_header_line(const unsigned char* buffer, size_t len, size_t*
...
+            LOG(AZ_LOG_TRACE, LOG_LINE, "%.*s Proxy-Authorization: ************** %.*s", (int)(proxyAuthStart  - text_line), text_line, (int)(strlen(text_line) - (proxyAuthEol - text_line)), proxyAuthEol);
+        }
+    }
+    else
+    {
+        int authPos = authStart - text_line;
+        int proxyAuthPos = proxyAuthStart - text_line;
+        if(authPos < proxyAuthPos)
+        {
+            LOG(AZ_LOG_TRACE, LOG_LINE, "%.*s Authorization: ************** %.*s Proxy-Authorization: ************** %.*s", (int)(authStart - text_line), text_line, (int)(proxyAuthStart - authEol), authEol,
```
[the remark this answers is not in the document]
Comment (JoseSanchez7) [reply src/uhttp.c]: Fixed.

File: src/uhttp.c [hunk 21]
```diff
@@ -307,9 +310,63 @@ static int process_header_line(const unsigned char* buffer, size_t len, size_t*
...
+(int)(proxyAuthEol - text_line),
+proxyAuthEol);
+        }
+        else
+        {
+            LOG(AZ_LOG_TRACE, LOG_LINE, "%.*s Proxy-Authorization: ************** %.*s Authorization: ************** %.*s", (int)(proxyAuthStart - text_line), text_line, (int)(authStart - proxyAuthEol), proxyAuthEol,(int)(authEol -
```
[the remark this answers is not in the document]
Comment (JoseSanchez7) [reply src/uhttp.c]: Fixed.

[the remark this answers is not in the document]
Comment (JoseSanchez7) [reply src/uhttp.c]: Fixed.

File: src/uhttp.c (hunk 14 above)
Comment (JoseSanchez7) [reply src/uhttp.c:314]: Fixed.

File: src/uhttp.c (hunk 15 above)
Comment (JoseSanchez7) [reply src/uhttp.c]: Good point! Fixed.

File: src/uhttp.c (hunk 16 above)
Comment (JoseSanchez7) [reply src/uhttp.c]: Good point! Fixed.

File: src/uhttp.c (hunk 12 above)
Comment (ewertons) [reply src/uhttp.c]: Just chiming in here, this function (log_text_line) is static, so it is not directly testable through unit tests. We can add tests calling the public functions that indirectly call log_text_line(), but we also do not unit-test logging in the SDK. So these unit tests would not assert anything related to the new logging, but can be used for a manual verification in the console that the expected logs are being printed.
Please share any ideas if you think otherwise.

File: src/uhttp.c [hunk 22]
```diff
@@ -307,9 +309,56 @@ static int process_header_line(const unsigned char* buffer, size_t len, size_t*
     return result;
 }
 
+static void log_text_line(HTTP_CLIENT_HANDLE_DATA* http_data, const char* text_line)
```
Comment (vaavva) [inline src/uhttp.c:312]: is http_data used in the function?

File: src/uhttp.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (vaavva) [reply src/uhttp.c]: unresolving as this comment is still applicable!

File: src/uhttp.c (hunk 22 above)
Comment (JoseSanchez7) [reply src/uhttp.c:312]: Good catch! Removing it.

File: src/uhttp.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (JoseSanchez7) [reply src/uhttp.c]: @vaavva Thanks for catching that, went ahead and fixed it!

**Review by vaavva — APPROVED** [review]

File: src/uhttp.c [hunk 23]
```diff
@@ -307,9 +309,56 @@ static int process_header_line(const unsigned char* buffer, size_t len, size_t*
...
+    }
+    if (authStart == NULL && proxyAuthStart == NULL)
+    {
+        LOG(AZ_LOG_TRACE, LOG_LINE, "%s", text_line);
+    }
+    else if (proxyAuthStart == NULL)
+    {
+        if (authEol != NULL)
+        {                                                                 
+            LOG(AZ_LOG_TRACE, LOG_LINE, "%.*s \r\nAuthorization: *** %.*s", (int)(authStart  - text_line), text_line, (int)(strlen(text_line) - (authEol - text_line)), authEol);
```
Comment (danewalton) [inline src/uhttp.c:335]: can we get these long log lines shorter with newlines for each of the variables?
```suggestion
            LOG(AZ_LOG_TRACE, LOG_LINE, "%.*s \r\nAuthorization: *** %.*s",
                                                                 (int)(authStart  - text_line), text_line,
                                                                 (int)(strlen(text_line) - (authEol - text_line)), authEol);
```

File: src/uhttp.c [hunk 24]
```diff
@@ -307,9 +309,56 @@ static int process_header_line(const unsigned char* buffer, size_t len, size_t*
...
+        }
+    }
+    else if (authStart == NULL)
+    {
+        if (proxyAuthEol != NULL)
+        {
+            LOG(AZ_LOG_TRACE, LOG_LINE, "%.*s \r\nProxy-Authorization: *** %.*s", (int)(proxyAuthStart  - text_line), text_line, (int)(strlen(text_line) - (proxyAuthEol - text_line)), proxyAuthEol);
```
Comment (danewalton) [inline src/uhttp.c:342]:
```suggestion
            LOG(AZ_LOG_TRACE, LOG_LINE, "%.*s \r\nProxy-Authorization: *** %.*s",
                                                     (int)(proxyAuthStart  - text_line), text_line,
                                                     (int)(strlen(text_line) - (proxyAuthEol - text_line)), proxyAuthEol);
```

File: src/uhttp.c [hunk 25]
```diff
@@ -307,9 +309,56 @@ static int process_header_line(const unsigned char* buffer, size_t len, size_t*
...
+        }
+    }
+    else
+    {
+        if(authStart < proxyAuthStart)
+        {
+            LOG(AZ_LOG_TRACE, LOG_LINE, "%.*s \r\nAuthorization: *** %.*s \r\nProxy-Authorization: *** %.*s", (int)(authStart - text_line), text_line, (int)(proxyAuthStart - authEol), authEol, (int)(proxyAuthEol - text_line), proxyAuthEol);
```
Comment (danewalton) [inline src/uhttp.c:349]:
```suggestion
            LOG(AZ_LOG_TRACE, LOG_LINE, "%.*s \r\nAuthorization: *** %.*s \r\nProxy-Authorization: *** %.*s",
                               (int)(authStart - text_line), text_line,
                               (int)(proxyAuthStart - authEol), authEol,
                               (int)(proxyAuthEol - text_line), proxyAuthEol);
```

File: src/uhttp.c [hunk 26]
```diff
@@ -307,9 +309,56 @@ static int process_header_line(const unsigned char* buffer, size_t len, size_t*
...
+        }
+        else
+        {
+            LOG(AZ_LOG_TRACE, LOG_LINE, "%.*s \r\nProxy-Authorization: *** %.*s \r\nAuthorization: *** %.*s", (int)(proxyAuthStart - text_line), text_line, (int)(authStart - proxyAuthEol), proxyAuthEol, (int)(authEol - text_line), authEol);
```
Comment (danewalton) [inline src/uhttp.c:353]:
```suggestion
            LOG(AZ_LOG_TRACE, LOG_LINE, "%.*s \r\nProxy-Authorization: *** %.*s \r\nAuthorization: *** %.*s",
            (int)(proxyAuthStart - text_line), text_line,
            (int)(authStart - proxyAuthEol), proxyAuthEol,
            (int)(authEol - text_line), authEol);
```

**Review by danewalton — APPROVED** [review]

File: src/uhttp.c [hunk 27]
```diff
@@ -32,6 +32,8 @@ static const char* HTTP_HOST = "Host";
 static const char* HTTP_CONTENT_LEN = "content-length";
 static const char* HTTP_TRANSFER_ENCODING = "transfer-encoding";
 static const char* HTTP_CRLF_VALUE = "\r\n";
+static const char* AUTH_HEADER = "\r\nAuthorization:";
+static const char* PROXY_AUTH_HEADER = "\r\nProxy-Authorization:";
```
[the remark this answers is not in the document]
Comment (JoseSanchez7) [reply src/uhttp.c]: Fixed. Thanks!
